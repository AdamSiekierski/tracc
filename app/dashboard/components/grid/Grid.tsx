import React, { ReactNode } from "react"

type GridProps = {
  children: ReactNode
  maxCols?: number
}

const Grid = ({ children, maxCols = 2 }: GridProps) => {
  return <section className={`grid grid-cols-${maxCols} gap-4 h-full`}>{children}</section>
}

export default Grid
