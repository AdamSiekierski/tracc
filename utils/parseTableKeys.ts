export function parseTableKeys<T>(data: T[]) {
  return [
    ...data.reduce((acc, obj) => {
      for (const key of Object.keys(obj)) {
        acc.add(key)
      }

      return acc
    }, new Set<string>([])),
  ]
}
