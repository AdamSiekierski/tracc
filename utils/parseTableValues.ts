export function parseTableValues<T>(data: T[], keys: string[]): string[][] {
  return data.reduce((acc: string[][], obj: T) => {
    const currentValues: string[] = []

    for (const key of keys) {
      currentValues.push(obj[key] || "")
    }

    return [...acc, currentValues]
  }, [])
}
