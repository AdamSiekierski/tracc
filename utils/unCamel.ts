export const unCamel = (str) => str.replace(/([A-Z])/g, " $1").replace(/^./, (s) => s.toUpperCase())
