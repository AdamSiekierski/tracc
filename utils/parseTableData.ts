import { parseTableKeys } from "./parseTableKeys"
import { parseTableValues } from "./parseTableValues"

export function parseTableData<T>(data: T[]): [string[], string[][]] {
  const keys = parseTableKeys<T>(data)
  const rows = parseTableValues<T>(data, keys)

  return [keys, rows]
}
