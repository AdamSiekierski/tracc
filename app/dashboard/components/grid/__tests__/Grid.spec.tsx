import React from "react"
import Grid from "../Grid"
import { render } from "test/utils"
import { cleanup } from "@testing-library/react"

describe("grid/Grid", () => {
  const MockComponent = () => <div data-testid="mock-component" />

  afterEach(cleanup)

  it("renders given elements correctly", () => {
    const { getAllByTestId, asFragment } = render(
      <Grid>
        <MockComponent />
        <MockComponent />
      </Grid>
    )

    const allElements = getAllByTestId("mock-component")

    expect(allElements).toHaveLength(2)
    expect(asFragment()).toMatchSnapshot()
  })

  it("respects the max number of columns", () => {
    const { asFragment } = render(
      <Grid maxCols={1}>
        <MockComponent />
        <MockComponent />
      </Grid>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
