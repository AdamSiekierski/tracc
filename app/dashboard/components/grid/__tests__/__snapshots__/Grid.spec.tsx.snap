// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid/Grid renders given elements correctly 1`] = `
<DocumentFragment>
  <section
    class="grid grid-cols-2 gap-4 h-full"
  >
    <div
      data-testid="mock-component"
    />
    <div
      data-testid="mock-component"
    />
  </section>
</DocumentFragment>
`;

exports[`grid/Grid respects the max number of columns 1`] = `
<DocumentFragment>
  <section
    class="grid grid-cols-1 gap-4 h-full"
  >
    <div
      data-testid="mock-component"
    />
    <div
      data-testid="mock-component"
    />
  </section>
</DocumentFragment>
`;
